python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,148 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Synchronous task scheduler."""
+
+from __future__ import annotations
+
+from collections.abc import MutableMapping
+from typing import TYPE_CHECKING, Any, TypeVar
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+    from typing import TypeAlias
+
+
+Key: TypeAlias = str | tuple[str, int]
+Graph: TypeAlias = MutableMapping[Key, Any]
+T_ = TypeVar("T_")
+
+
+# NOTE: This is a slimmed-down version of the single-threaded
+# (synchronous) scheduler in `dask.core`.
+#
+# Key Differences:
+# * We do not allow a task to contain a list of key names.
+#   Keys must be distinct elements of the task.
+
+
+def istask(x: Any) -> bool:
+    """Check if x is a callable task."""
+    return bool(type(x) is tuple and x and callable(x[0]))
+
+
+def _execute_task(arg: Any, cache: MutableMapping) -> Any:
+    """Execute a compute task."""
+    if istask(arg):
+        return arg[0](*(_execute_task(a, cache) for a in arg[1:]))
+    else:
+        try:
+            return cache.get(arg, arg)
+        except TypeError:  # Unhashable task argument
+            return arg
+
+
+def required_keys(key: Key, graph: Graph) -> list[Key]:
+    """
+    Return the dependencies to extract a key from the graph.
+
+    Parameters
+    ----------
+    key
+        Root key we want to extract.
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    keys = []
+    tasks = [graph[key]]
+    while tasks:
+        work = []
+        for w in tasks:
+            if istask(w):
+                work.extend(w[1:])
+            else:
+                try:
+                    if w in graph:
+                        keys.append(w)
+                except TypeError:  # not hashable
+                    pass
+        tasks = work
+    return keys
```

Comment:
I think we never have nested tasks, so this could be much simpler?

```
def required_keys(key, graph):
    maybe_task = graph[key]
    if istask(maybe_task):
        return [k for k in maybe_task[1:] if k in graph]
    return []
```
?

Reply:
>I think we never have nested tasks

Actually we do still have nested tasks in Join :/ (I'll see how much it hurts performance to remove that, since that's probably breaking rmpf spilling)

Also, `graph[key]` can return another key, and the elements of a task may not be hashable (so we need to catch a `TypeError` for the `if k in graph` check).

Reply:
I was able to remove nested tasks and simplify this a bit.